mirascope/base/extractors.py

```diff
@@ -194,6 +211,181 @@ class TempCall(call_type):  # type: ignore
+            call_type: The type of call to use for extraction. This enables shared code
+                across various model providers that have slight variations but the same
+                internal interfaces.
+            tool_type: The type of tool to use for extraction. This enables shared code
+                across various model providers that have slight variations but the same
+                internal interfaces.
+            tool_stream_type: The type of tool stream to use for streaming tools. This
+                enables shared code across various model providers that have slight
+                variations but the same internal interfaces.
+            retries: The number of call attempts to make on `ValidationError` before
+                giving up and throwing the error to the user.
+            **kwargs: Additional keyword arguments.
+
+        Yields:
+            An instance of the partial `extract_schema` with it's available fields
+            populated.
+
+        Raises:
+            AttributeError: if there is no tool in the call creation.
+            ValidationError: if the schema cannot be instantiated from the completion.
+        """
+        kwargs, return_tool = self._setup(tool_type, kwargs)
+
+        class TempCall(call_type):  # type: ignore
+            prompt_template = self.prompt_template
+
+            base_url = self.base_url
+            api_key = self.api_key
+            call_params = self.call_params
+
+            model_config = ConfigDict(extra="allow")
+
+        setattr(TempCall, "messages", self.messages)
+        stream = TempCall(**self.model_dump(exclude={"extract_schema"})).stream_async(
+            **kwargs
+        )
+        tool_stream = tool_stream_type.from_async_stream(stream, allow_partial=True)
+        try:
+            yielded = False
+            async for partial_tool in tool_stream:
+                extracted_schema = self._extract_schema(
+                    partial_tool, self.extract_schema, return_tool, response=None
+                )
+                if extracted_schema is None:
+                    break
+                yielded = True
+                yield extracted_schema
+
+            if not yielded:
+                raise AttributeError("No tool found in the completion.")
+        except (AttributeError, ValueError, ValidationError) as e:
+            if retries > 0:
+                logging.info(f"Retrying due to exception: {e}")
+                # TODO: include failure in retry prompt.
+                async for partial_tool in self._stream_async(
+                    call_type, tool_type, tool_stream_type, retries - 1, **kwargs
+                ):
+                    yield partial_tool
+            raise  # re-raise if we have no retries left
+
+    def _extract_schema(
+        self,
+        tool: Optional[BaseToolT],
+        schema: ExtractedType,
+        return_tool: bool,
+        response: Optional[Any],
+    ) -> Optional[ExtractedTypeT]:
```

Comment:
should the return type be `Optional[Union[ExtractedTypeT, BaseToolT]]`

Reply:
Technically BaseToolT is part of ExtractedTypeT but mypy is annoying